### ISSUE TITLE
How do I handle the exception - System.AggregateException A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread.

### ISSUE THREAD

**Issue by marafiq** [Azure/azure-signalr#1928]
### Describe the bug
Getting the below exception when sending message to client using Groups.  
How do I avoid  this or gracefully handle this?
Like below

```
public static void Notify(long Id, params string[] keys)
  {
      GetHubContext<CountsHub>()
          .Clients
          .Group(GroupKey.FormatWith(Id))
          .notify(GetCounts(Id, default, keys));
  }
```

### Exceptions (if any)
```
System.InvalidOperationException: No service connection found when sending message
  File "AzureTransport.cs", line 106, in Task AzureTransport.Send(object value)
    throw new InvalidOperationException("No service connection found when sending message");
  File "/_/src/Microsoft.AspNet.SignalR.Core/Hubs/HubDispatcher.cs", line 235, col 21, in Task HubDispatcher.InvokeHubPipeline(IHub hub, IJsonValue[] parameterValues, MethodDescriptor methodDescriptor, HubRequest hubRequest, StateChangeTracker tracker)+(Task<object> task) => { } [1]
  File "/_/src/Microsoft.AspNet.SignalR.Core/TaskAsyncHelper.cs", line 1006, col 17, in TResult TaskAsyncHelper.RunWithPreservedCulture<T1, T2, TResult>(CulturePair preservedCulture, Func<T1, T2, TResult> func, T1 arg1, T2 arg2)
  File "/_/src/Microsoft.AspNet.SignalR.Core/TaskAsyncHelper.cs", line 1017, col 13, in TResult TaskAsyncHelper.RunWithPreservedCulture<T, TResult>(CulturePair preservedCulture, Func<T, TResult> func, T arg)
  File "/_/src/Microsoft.AspNet.SignalR.Core/TaskAsyncHelper.cs", line 1071, col 43, in Task TaskAsyncHelper.ContinueWithPreservedCulture(Task task, Action<Task> continuationAction, TaskContinuationOptions continuationOptions)+(Task<T> t) => { }
  ?, in void ContinuationResultTaskFromResultTask<TAntecedentResult, TResult>.InnerInvoke()
  ?, in void Task.Execute()
```
### Further technical details
- Your Azure SignalR SDK version - 1.21.4
- ASP.NET 4.8
- Your SignalR Client SDK version - 2.4.1

**Comment by vicancy** [comment]
Need to fix, currently the code throws at https://github.com/Azure/azure-signalr/blob/89f393086b829b0fe8426aaf6ad4b664c6aff5fc/src/Microsoft.Azure.SignalR.AspNet/Transports/AzureTransport.cs#L115

**Comment by marafiq** [comment]
@vicancy After updating to 2.4.3 the rate of exception has increased, getting 2x exceptions compared to previous version 2.4.1. 
Any workaround?

On some days exception happens more frequently when there are fewer active users. 

![Image](https://github.com/user-attachments/assets/697020fb-5f33-4df7-a146-eb0b1ba2f457)

**Comment by vicancy** [comment]
Hi @marafiq could you share with me your resource name (send me through `lianwei<at>microsoft.com`), let me check if I can get any hints from the service logs.

**Comment by marafiq** [comment]
@vicancy thank you. Sent you the information via email.

**Comment by vicancy** [comment]
Such exception happens when server connections are dropped and messages failed to be sent out.

With the resource provided, the resource metrics shows that the exception increasing is probably caused by high server load and server connections closed by "PingTimeout", suggest increasing the resource Unit to see if the issue is mitigated.

**Comment by marafiq** [comment]
Thanks @vicancy for looking into it. After scaling the instance of SignalR Service so that usage stays under 70% these exception are still happening. 

>With the resource provided, the resource metrics shows that the exception increasing is probably caused by high server load and server connections closed by "PingTimeout"

I do not know if these are related or not, but I am looking at the process memory dump and noticing high fragmentation in Gen0 caused by byte[] rooted in SignalR websocket.

<details><summary>Memory Dump Pinned Byte Arrays coming from SignalR in Gen0</summary>
<p>

634 MB free space in GEN0 

1. ![Image](https://github.com/user-attachments/assets/f8a62748-3058-4102-9b05-17d5b2d8bfe4)


2.  Group by Byte[] - 150 and 97 objects are pinned 

![Image](https://github.com/user-attachments/assets/3fe552b3-bfa0-4f0d-9c1e-efea7bcbc779)

3. When looking into 150 objects of byte[] - each array is of size 49316 and is pinned by SignalR

![Image](https://github.com/user-attachments/assets/a1c04e24-5ef7-44e6-8d32-1bc13f6d46df)

</p>
</details> 

I am suspecting that ping timeouts are failing perhaps due to GC pressure at times, most importantly whey there are so many pinned arrays in Gen0, is it expected? If not what can I do differently.

**Comment by marafiq** [comment]
With respect to exceptions happening at similar rate - last 3 days. Singal is using 4 units in Premium plan

![Image](https://github.com/user-attachments/assets/218fb276-531f-4da4-b851-097d9b68d420)

And after the scale - Blue line is server load, and red is connection quota

![Image](https://github.com/user-attachments/assets/f3b7f3cf-a575-43ee-9ee7-07580c593069)

**Comment by marafiq** [comment]
Interestingly all those bytes array contains similar pay load of bytes, I convert those to string. Is it due continuously checking signal r service capacity?  
`???status?0?clientcount?43?servercount?2643?capacity?30000`

Is it somehow related to `EndpointMetrics`

I would like to note that we are using SignalR with 2.4.3 in .NET 4.8. The only thing which might be little different than normal usage is that we use same SignalR instance by logically seperating connection by mapping `app.MapAzureSignalR($"{appName}"` with app name. Thus there will be many application instances with diff apps connected to it.

**Comment by vicancy** [comment]
Hi @marafiq, what is the memory/CPU usage of your app server instance? If the issue happens after upgrading to 2.4.3 SignalR, it might not be related to these G0 arrays, they are not related to 2.4.3 SignalR. Maybe SignalR 2.4.3 assumes more memory and adds memory pressure.

Under the hood, an `{appName}` introduce in `(1+N)*[InitialHubServerConnectionCount,MaxHubServerConnectionCount])` server connections, `N` is the number of hubs, and introduce in `(1+N)` such ping messages. If you have `M` different app names, there would be `M*(1+N)` ping checks, so I think these 150 pinned array objects are the messages received from the service.  (It is indeed something to improve in the SignalR service side, to reduce the frequency of sending such redundant info)

What is the current `InitialHubServerConnectionCount` and `MaxHubServerConnectionCount` set in your ServiceOptions(In old SDK the option is called ServerConnectionCount)? If your client connections and traffic is not large, maybe decreasing the count from default `5` to a smaller number `3` might help decrease the memory usage.

**Comment by marafiq** [comment]
Thank you @vicancy 

Let me try to answer some of questions asked above. I think overall memory usage (generally memory pressure) and `invalid opertion` exceptions are interlinked. If you think these are two separate issues, should I open another issue to fix the memory pinning problems.

> Is upgrade to 2.4.3 a problem?

Not entirely true. The number of errors has increased since upgrade, but these errors has been a problem from long time in diff versions. since I logged the error last year. Or its simply because our users have increase over time.

> InitialHubServerConnectionCount and Max connections?

I am using default which is 5+1 initial connections per hub. We have 20 hubs. I have tried decreasing this number which introduce another problem of 503 errors. Since connections are scaled up or down by the SDK thus starting low would not really help much when the usage is high.

> Resource Usage in App Service?

We are using Premium v3 P2V3. 

<details><summary>Memory & CPU Usage are Healthy</summary>
<p>

![Image](https://github.com/user-attachments/assets/7fcd54d3-e1e4-498d-9a1a-1f86423095c7)

![Image](https://github.com/user-attachments/assets/6c4dcd01-0d78-4734-b528-b9f5ab3eb6ba)

</p>
</details> 

The memory usage will certainly be better if there is less pinning. We are using high memory instance because sometime memory can jump around really bad, and fragmentation is big part of it.


> 150 pinned arrays - Loosing 600 MB due to Fragmentation in GEN0

Yes, I understand the number because we have many hubs. Reducing the number of hubs is one option. But pinning of these arrays does not help with overall app stability. Is it something which can be solved by using arrays from pool in the SDK itself.

But why these arrays are of size 49316 for end point metrics? Those arrays do not seem to use for messaging at all based on their content. 
Another set of pinned arrays are created for each web socket, should these be using buffer from memory pool?

<details><summary>97 Pinned Arrays in GEN0</summary>
<p>

![Image](https://github.com/user-attachments/assets/b231cebe-4dc6-470c-a25b-9a4beb418a52)

![Image](https://github.com/user-attachments/assets/e11cd602-2d17-4022-b1f6-2689502fc4ae)

</p>
</details>

**Comment by vicancy** [comment]
Thanks for analyzing the issue @marafiq, yes I agree with you that the memory issue is the root issue. It might be related to this different logic in WebSocketsTransport. Please allow me to look into the issue to see how to improve that.
https://github.com/Azure/azure-signalr/blob/ec579cea80cd3000416aaf28ee463fc03f26b01c/src/Microsoft.Azure.SignalR.Common/ServiceConnections/Internal/WebSocketsTransport.cs#L293

> I am using default which is 5+1 initial connections per hub. We have 20 hubs. I have tried decreasing this number which introduce another problem of 503 errors. Since connections are scaled up or down by the SDK thus starting low would not really help much when the usage is high.

For this, setting `MaxHubServerConnectionCount` controls the max count the connections could scale up to.

**Comment by vicancy** [comment]
Some findings:
For netframework, the underlying WebSocket connection is using [ClientWebSocket](https://github.com/Microsoft/referencesource/blob/master/System/net/System/Net/WebSockets/ClientWebSocket.cs#L169), in which the default `SendBufferSize` and `ReceiveBufferSize` is `16*1024`, every websocket [creates an internal byte[]](https://github.com/microsoft/referencesource/blob/51cf7850defa8a17d815b4700b67116e3fa283c2/System/net/System/Net/WebSockets/WebSocketBuffer.cs#L665) using `2 * receiveBufferSize + nativeSendBufferSize + NativeOverheadBufferSize + s_PropertyBufferSize`, that matches _**these arrays are of size 49316**_

Every websocket connection creates a pinned array buffer, so decreasing the hub number or server connection count might help. Does `InitialHubServerConnectionCount=MaxHubServerConnectionCount=3` work? what is the error detail for 503? 

I am doing some test with different send/receive buffer size, with no traffic but 200 server connections connected, will update later with findings.

**Comment by marafiq** [comment]
@vicancy Thank you. 

> For netframework, the underlying WebSocket connection is using [ClientWebSocket](https://github.com/Microsoft/referencesource/blob/master/System/net/System/Net/WebSockets/ClientWebSocket.cs#L169), in which the default SendBufferSize and ReceiveBufferSize is 16*1024, every websocket [creates an internal byte[]](https://github.com/microsoft/referencesource/blob/51cf7850defa8a17d815b4700b67116e3fa283c2/System/net/System/Net/WebSockets/WebSocketBuffer.cs#L665) using 2 * receiveBufferSize + nativeSendBufferSize + NativeOverheadBufferSize + s_PropertyBufferSize, that matches these arrays are of size 49316

Ah makes sense. And array of length `byte[14616]` are directly result of WebSocketTransport.

> Every websocket connection creates a pinned array buffer, so decreasing the hub number or server connection count might help. Does InitialHubServerConnectionCount=MaxHubServerConnectionCount=3 work? what is the error detail for 503?

I have tried with min=3 to max=5 connections.  Its throws 503 - `SingalR` Service not available. I will give it another go in next release. But in my view it will not make a difference because:

From the memory dump I know some hubs are using 10+ connections, and some are using 3-10 connections. So it seems like `SignalR` is doing good job on scaling up & down.  Even if I start with low connections, SignalR is going to scale connections based on load. Thus, pinning is back.
Actually, pinning can be worse given the initial count is low, the connections created later in time may end up at heap location which will cause more fragmentation vs starting with default count, at least allocated closed to each other (in theory not that they are)

In fact, initial and max hub server connections should be hub specific rather than at app level, but that's feature request. There is strong use case to it. 

> I am doing some test with different send/receive buffer size, with no traffic but 200 server connections connected, will update later with findings.

I appreciate it. Looking forward to no pinning in Gen0. 

If this helps, my application message sizes are not large payload, its most 3-6 properties which carries count of diff things.

**Comment by vicancy** [comment]
> In fact, initial and max hub server connections should be hub specific rather than at app level, but that's feature request. There is strong use case to it.

yeah this makes sense. 

> Even if I start with low connections, SignalR is going to scale connections based on load. Thus, pinning is back.

SignalR service tries to rebalance the traffic and connection distribution, if your traffic is low, max=5 sounds enough. if max is set to 5, SignalR will not scale up connections. `SingalR Service not available.` happens when all the hub server connections are dropped. 

> Looking forward to no pinning in Gen0.

In theory, if a byte[] buffer is larger than 85KB, it is allocated on the Large Object Heap (LOH), and could reduce the G0 GC pressure. I did several rounds of tests, but I didn't see much difference in CPU usage when setting websocket buffer to `2K`/default/`30K` with 200 server connections connected running for 10 mins. With **Server GC mode**, the GC frequency is low. CPU usage profiling shows the CPU time is spent in WebSocket sending and receiving.

**Comment by marafiq** [comment]
> SignalR service tries to rebalance the traffic and connection distribution, if your traffic is low, max=5 sounds enough. if max is set to 5, SignalR will not scale up connections. SingalR Service not available. happens when all the hub server connections are dropped.

I will give it a try and update. 

> In theory, if a byte[] buffer is larger than 85KB, it is allocated on the Large Object Heap (LOH), and could reduce the G0 GC pressure. I did several rounds of tests, but I didn't see much difference in CPU usage when setting websocket buffer to 2K/default/30K with 200 server connections connected running for 10 mins. With Server GC mode, the GC frequency is low. CPU usage profiling shows the CPU time is spent in WebSocket sending and receiving.

Can not we use the buffer from ArrayPool or build pool of sockets so that pinning can be avoided. If not then allocating it in LOH may help in my case because I have very busy Gen0 due to MVC allocations with thousands of routes, and pinning in Gen0 is big issue.

**Comment by vicancy** [comment]
Coud you try this preview version with the underlying ClientWebSocketOptions  exposed?
`dotnet add package Microsoft.Azure.SignalR.AspNet --version 1.30.3-preview1-11320 --source https://www.myget.org/F/azure-signalr-dev/api/v3/index.json`


```cs
app.RunAzureSignalR(..., o => o.ConfigureServiceConnectionWebSocketOptions = i => i.SetBuffer(bufferSize, bufferSize))
```

**Comment by marafiq** [comment]
@vicancy Thanks I will give it a try.

**Comment by marafiq** [comment]
@vicancy It seems like the preview version is using hard reference for Microsoft.Extensions.Configuration.Abstractions 2.1.1.0 so my project facing assembly redirect issue because I use version 8.0.0.0 ATM. And I have installed the 8.0.0.0 directly into project.

![Image](https://github.com/user-attachments/assets/2cbc2abe-4bbe-4ad8-bdd9-66ab4807f0ea)

![Image](https://github.com/user-attachments/assets/29c3dbaa-5b1d-44da-b4b8-ebaf59a32151)

![Image](https://github.com/user-attachments/assets/e61a820a-70dc-48a3-a977-c4383335b518)

**Comment by vicancy** [comment]
the new version upgraded the package versions to 8.0.x. could you share with me the error details?

**Comment by marafiq** [comment]
Sorry updated the last comment. Its only package I updated and above errors started happening.
> the new version upgraded the package versions to 8.0.x. could you share with me the error details?

The downloaded package still reference 2.1.10 and same is downloaded

![Image](https://github.com/user-attachments/assets/6971c222-67db-486e-bf2d-6af6c9610a77)

Note: I tried both preview versions ending with 20 and 21

**Comment by marafiq** [comment]
update: I can only make it work by copying the 8.0.0.0 dll to bin folder manually and binding redirect pointing to 8.0.0.0 because otherwise visual studio will generate binding redirect for 2.1.0.0. But it wont work in publish mode.
The install added below line to project file

![Image](https://github.com/user-attachments/assets/d5505ab7-678e-4c9a-93ce-73e9526cf297)

**Comment by marafiq** [comment]
More update on it, If i manually revert the project file reference pointing to 8.0.0.0 - to actual extension package the visual studio warnings go away and project loads. But there is definitely something wrong with references.

**Comment by marafiq** [comment]
Memory dump from local, these 16416 size buffers are still lurking in Gen 0 - to be sure I checked the service point and its SignalR pinning 
and I am using this `options.ConfigureServiceConnectionWebSocketOptions = i => i.SetBuffer(8501, 8501);`

![Image](https://github.com/user-attachments/assets/3a769960-e350-47d1-9db4-ae89e0938309)

**Comment by marafiq** [comment]
I have tried using size `9500` to be sure that it lands in LOH but I do not see anything in LOH, I was using version `1.30.3-preview1-11321` but now I used `1.30.3-preview1-11320`

The only diff is the size of array pinned in Gen0 because I do not see any byte array in LOH for the SignalR. Having said that there seems to be little improvement in fragmentation in Gen0 but does not look promising. 
Can you please explain, what should I expect to see in memory dump when I provide buffer of size 9500 for both buffers? @vicancy 

![Image](https://github.com/user-attachments/assets/ab1a5905-5f8f-4085-b465-cf8ea212b2ba)

**Comment by marafiq** [comment]
Damn it's 85000 not 8500

**Comment by marafiq** [comment]
Ok, its combined size. I used the max allowed size in buffer. `options.ConfigureServiceConnectionWebSocketOptions = i => i.SetBuffer(65536, 65536);`
Now I see these allocated in LOH 

![Image](https://github.com/user-attachments/assets/1e8ded5e-2268-4966-96e3-2899f2a7ed73)

But not sure why I am still seeing the below byte arrays?

![Image](https://github.com/user-attachments/assets/1c85f843-efe1-4b19-aff8-ade526943eab)

Do I need to use this overload `SetBuffer(int receiveBufferSize, int sendBufferSize, ArraySegment<byte> buffer)`

**Comment by marafiq** [comment]
@vicancy did you get a chance to take a look?

**Comment by vicancy** [comment]
looks like these bytes are preserved by SslStream and I am afraid we could do little about these bytes https://github.com/microsoft/referencesource/blob/51cf7850defa8a17d815b4700b67116e3fa283c2/System/net/System/Net/SecureProtocols/_SslStream.cs#L42

Reducing the concurrent server connection count sounds the way to mitigate the issue.

**Comment by marafiq** [comment]
@vicancy Thanks. 

> Reducing the concurrent server connection count sounds the way to mitigate the issue.
Essentially, reduce the number of hubs in my scenario. 

Additionally, would you mind providing hooks to configure connections per hub.

**Comment by vicancy** [comment]
> providing hooks to configure connections per hub.

Would you mind creating a feature request issue for this?

**Comment by marafiq** [comment]
@vicancy Will do so in near future. But I can reproduce the bug now.

Say you have hub client method `Clients.Group(GetFormattedGroupKey(categoryId, User)).notify(GetCountFromDatabase())` and you are on a page `/Page1` but query took more than 20 seconds, after the hub connection & ping, user navigate to another page `/Page2` and stayed there. By the time query will return connection is already closed thus `no connection found` 

I think either this exception should be optional to throw or connection id should be attached after notify call somehow.

**Comment by marafiq** [comment]
@vicancy 
> {"time":"2025-08-09T19:28:57.1327894Z","logLevel":"Warning","eventName":"NoConnectionInGroup","message":"Group message [957288669146775562] was not received because there's no active connection in group [hg-xHub.xGroup_3].","exception":null,"messageTracingId":"957288669146775562"}

So, it turns out when users move away quickly from the page, and its only user in that group. its when this exception is being thrown. 

Is there a workaround that or at least can this be made optional behavior or at least do not throw exception when group is empty correctly. We have been dealing with the noise from long time now as seen below. At least this settle the root cause (would like to believe it after all we looked at). Would appreciate any help

<img width="1711" height="302" alt="Image" src="https://github.com/user-attachments/assets/23e5f09f-f5fd-412e-981e-a08119df00b6" />

**Comment by marafiq** [comment]
Hi @vicancy 
I updated to the latest version, but it is causing new issue, where legit message is not being sent when the connection is active, and ping is being received in console as well. See the logs below which shows subscription succeeded but right after connection is not found but when I refreshed the page, same connection id was disconnected. 




Time | Log Level | Event Name | Message | Hub | Connection ID
-- | -- | -- | -- | -- | --
2025/10/17 18:44:35:170 | Information | RequestProcessed | Request processed. |   | t5Ayme5MebXwQ4j6M8rpjQVmy0Pw802
2025/10/17 18:44:35:614 | Information | RequestProcessed | Request processed. |   | t5Ayme5MebXwQ4j6M8rpjQVmy0Pw802
2025/10/17 18:44:35:619 | Information | ConnectionEnded | Connection ended. Reason: | priorityprospectshub;prospecttaskshub;stafftaskshub | t5Ayme5MebXwQ4j6M8rpjQVmy0Pw802
2025/10/17 18:44:36:246 | Information | RequestProcessed | Request processed. |   |  
2025/10/17 18:44:36:315 | Information | RequestProcessed | Request processed. |   | H9qBr3GzC9HpG-OQwelXqwRqotBQ802
2025/10/17 18:44:36:555 | Information | RequestStarted | Request started. |   | H9qBr3GzC9HpG-OQwelXqwRqotBQ802
2025/10/17 18:44:36:699 | Information | RequestProcessed | Request processed. |   | H9qBr3GzC9HpG-OQwelXqwRqotBQ802
2025/10/17 18:44:36:700 | Information | ConnectionStarted | Connection started | priorityprospectshub;prospecttaskshub;stafftaskshub | H9qBr3GzC9HpG-OQwelXqwRqotBQ802
2025/10/17 18:44:36:784 | Information | AddedConnectionToGroup | Added connection H9qBr3GzC9HpG-OQwelXqwRqotBQ802 to group   hg-PriorityProspectsHub.PriorityProspects:76 via message [61061421198213453]   successfully. |   | H9qBr3GzC9HpG-OQwelXqwRqotBQ802
2025/10/17 18:44:36:784 | Information | AddedConnectionToGroup | Added connection H9qBr3GzC9HpG-OQwelXqwRqotBQ802 to group   hg-ProspectTasksHub.ProspectTasks:76 via message [61061421198213454]   successfully. |   | H9qBr3GzC9HpG-OQwelXqwRqotBQ802
2025/10/17 18:44:37:650 | Warning | NoConnectionInGroup | Group message [61061421198213455] was not received because there's no   active connection in group [hg-ProspectTasksHub.ProspectTasks:78]. |   |  
2025/10/17 18:44:37:915 | Warning | NoConnectionInGroup | Group message [61061421198213456] was not received because there's no   active connection in group [hg-PriorityProspectsHub.PriorityProspects:78]. |   |  
2025/10/17 18:44:50:110 | Information | RequestProcessed | Request processed. |   | H9qBr3GzC9HpG-OQwelXqwRqotBQ802
2025/10/17 18:44:50:170 | Information | RequestProcessed | Request processed. |   | H9qBr3GzC9HpG-OQwelXqwRqotBQ802
2025/10/17 18:44:50:177 | Information | ConnectionEnded | Connection ended. Reason: | priorityprospectshub;prospecttaskshub;stafftaskshub | H9qBr3GzC9HpG-OQwelXqwRqotBQ802

**Comment by marafiq** [comment]
@vicancy Can you please help, is it possible that this issue is introduced in new versions only?

**Comment by vicancy** [comment]
Hi @marafiq is the error only happening with the new version? from the logs I see connection was added to group `hg-PriorityProspectsHub.PriorityProspects:76` while NoConnectionInGroup warns for group `hg-PriorityProspectsHub.PriorityProspects:78` which looks expected.

**Comment by marafiq** [comment]
@vicancy I apologize. You are right, it was logical bug, subscription and sending was out of sync in the app.
